library/alloc/src/sync/tests.rs

```diff
@@ -101,6 +101,45 @@ fn try_unwrap() {
     assert_eq!(Arc::try_unwrap(x), Ok(5));
 }
 
+#[test]
+fn unwrap_or_drop() {
+    // FIXME: Is doing this kind of loop reasonable? I tested `Arc::try_unwrap(x).ok()`
+    // and it makes this kind of assertion fail in roughly every second run somewhere
+    // between 1000 and 5000 iterations; I feel like doing a single iteration is too
+    // unlikely to catch anything interesting but doing too many is way too slow
+    // for a test that wouldn't ever fail for any reasonable implementation
+
+    for _ in 0..100
+    // ^ increase chances of hitting uncommon race conditions
+    {
+        use std::sync::Arc;
+        let x = Arc::new(3);
+        let y = Arc::clone(&x);
+        let r_thread = std::thread::spawn(|| Arc::try_unwrap(x).ok());
+        let s_thread = std::thread::spawn(|| Arc::try_unwrap(y).ok());
```

Comment:
Spawning a thread takes a while, so I would almost expect both `try_unwraps` to never run at the same time. Maybe try spawning both threads but then immediately blocking them to synchronize them. Then after both threads are spawned, in the main thread, you somehow signal to both threads that they may start now. I am not sure how best to do that, though. `RwLock` or rendezvous channel maybe?

But yeah, in any case, that's a really tricky test to write. Not sure what's best here.

Reply:
I tested and the test does indeed fail more often than one would think (on my machine) if `try_unwrap(x).ok()` is used (at least on my machine). Which makes me notice... I totally forgot replacing it with `unwrap_or_drop` again after testing that these failures can happen!! Damn... this means also means that it _didn’t_ fail once on the CI. But the hope was basically only that if `unwrap_or_drop` changed in the future and became broken (i.e. lost its guarantees of being atomic) then the test would fail for a lot of people (even if not for everyone) and the error would be noticed eventually.